analyses/cell-type-neuroblastoma-04/scripts/03c_run-scanvi-label-transfer.py

```diff
@@ -0,0 +1,160 @@
+            file=sys.stderr,
+        )
+        arg_error = True
+    if not arg.reference_scanvi_model_dir.is_dir():
+        print(
+            f"The provided reference scANVI model could not be found at: {arg.reference_scanvi_model_dir}.",
+            file=sys.stderr,
+        )
+        arg_error = True
+
+    # Set up testing and accelerator settings
+    if arg.testing:
+        # limit max_epochs for faster runtime and ensure CPU
+        common_train_kwargs = {"max_epochs": 5, "accelerator": "cpu"}
+    else:
+        accelerator = "gpu" if torch.cuda.is_available() else "cpu"
+        # don't use max_epochs; let scvi pick the heuristic
+        common_train_kwargs = {"accelerator": accelerator}
+
+    # Read and check that query object has expected columns
+    query = anndata.read_h5ad(arg.query_file)
+    expected_columns = [BATCH_KEY, CELL_ID_KEY] + COVARIATE_KEYS
+
+    if not set(expected_columns).issubset(query.obs.columns):
+        print(
+            f"The query AnnData object is missing one or more expected columns: {set(expected_columns).difference(query.obs.columns)}.",
+            file=sys.stderr,
+        )
+        arg_error = True
+
+    # Exit if error(s)
+    if arg_error:
+        sys.exit(1)
+
+    # Set seed for reproducibility
+    # torch commands are ignored if GPU not present
+    scvi.settings.seed = arg.seed  # inherited by numpy and torch
+    torch.cuda.manual_seed(arg.seed)
+    torch.cuda.manual_seed_all(arg.seed)
+
+    # Load the trained scANVI model for label transfer
+    scanvi_model = scvi.model.SCANVI.load(arg.reference_scanvi_model_dir)
+
+    # Ensure anndata is using a sparse matrix for faster processing
+    query.X = csr_matrix(query.X)
+
+    ################################################
+    # Incorporate query data into the scANVI model #
+    ################################################
+
+    # Prepare query data for training
+    scvi.model.SCANVI.prepare_query_anndata(query, scanvi_model)
+    scanvi_query = scvi.model.SCANVI.load_query_data(query, scanvi_model)
+
+    # Train model and get latent dimensions, cell type predictions
+    scanvi_query.train(
+        **common_train_kwargs,
+        # additional scArches parameters
+        plan_kwargs={"weight_decay": 0.0},
+        check_val_every_n_epoch=5,
+    )
+    query.obsm[SCANVI_LATENT_KEY] = scanvi_query.get_latent_representation()
+    query.obs[SCANVI_PREDICTIONS_KEY] = scanvi_query.predict()
+
+    ################################################
+    ################ Export objects ################
+    ################################################
+
+    # Export the query-trained scANVI model
+    scanvi_query.save(arg.query_scanvi_model_dir, save_anndata=True, overwrite=True)
+
+    # Extract data frame of latent representation and predictions
+    latent_df = pd.DataFrame(query.obsm[SCANVI_LATENT_KEY])
+    latent_df = latent_df.rename(columns=lambda x: SCANVI_LATENT_KEY + "_" + str(x))
```

Comment:
With your love of `glue`, I would have expected you to use an fstring here!
```suggestion
    latent_df = latent_df.rename(columns=lambda x: f"{SCANVI_LATENT_KEY}_{x}")
```
(I don't think the `str()` is needed with an fstring?)